cpp/src/arrow/extension/fixed_shape_tensor.cc

```diff
@@ -293,40 +335,50 @@ const Result<std::shared_ptr<Tensor>> FixedShapeTensorArray::ToTensor() const {
   // To convert an array of n dimensional tensors to a n+1 dimensional tensor we
   // interpret the array's length as the first dimension the new tensor.
 
-  auto ext_arr = std::static_pointer_cast<FixedSizeListArray>(this->storage());
-  auto ext_type = internal::checked_pointer_cast<FixedShapeTensorType>(this->type());
-  ARROW_RETURN_IF(!is_fixed_width(*ext_arr->value_type()),
-                  Status::Invalid(ext_arr->value_type()->ToString(),
-                                  " is not valid data type for a tensor"));
-  auto permutation = ext_type->permutation();
+  const auto storage_array =
+      internal::checked_pointer_cast<FixedSizeListArray>(this->storage());
+  ARROW_ASSIGN_OR_RAISE(const auto flattened_storage_array, storage_array->Flatten());
```

Comment:
It might be better to move this back after the validation of the type (and add a validation that there are no nulls), because `Flatten()` can incur a copy of the data, and then you would unnecessarily already do a copy when the next type check would fail.

Reply:
Good point. Moved.